epoch.go

```diff
@@ -378,11 +376,35 @@ func (e *Epoch) handleFinalizationCertificateMessage(message *FinalizationCertif
 			zap.Stringer("NodeID", from))
 		return nil
 	}
+
+	round, exists := e.rounds[message.Finalization.Round]
+	if !exists {
+		e.Logger.Debug("Received finalization certificate for a future round", zap.Uint64("round", message.Finalization.Round))
+		e.collectFutureFinalizationCertificates(message)
+		return nil
+	}
+
+	if round.fCert != nil {
+		e.Logger.Debug("Received finalization for an already finalized round", zap.Uint64("round", message.Finalization.Round))
+		return nil
+	}
+
 	round.fCert = message
 
 	return e.persistFinalizationCertificate(*message)
 }
 
+func (e *Epoch) collectFutureFinalizationCertificates(fCert *FinalizationCertificate) {
+	fCertRound := fCert.Finalization.Round
+	endSeq := math.Min(float64(fCertRound), float64(e.maxRoundWindow+e.round))
+	if e.latestRoundKnown.num >= uint64(endSeq) {
+		// we have already sent out messages collecting future finalization certificates
+		return
+	}
+	e.setLastReceivedFCertSeq(fCertRound)
+	e.sendFutureCertficatesRequests(e.round, uint64(endSeq), e.Comm)
```

Comment:
`e.round` and `endSeq` are different units. Since you may progress the round without progressing the sequence, it looks incorrect to me to mix them together. 

e.round should probably be the next sequence to what we're currently at.